packages/core/lib/segments/attributes/subsegment.js

```diff
@@ -123,6 +123,15 @@ Subsegment.prototype.addAnnotation = function(key, value) {
   } else if (typeof key !== 'string') {
     throw new Error('Failed to add annotation key: ' + key + ' value: ' + value + ' to subsegment ' +
       this.name + '. Key must be of type string.');
+  } else if (typeof key === 'string'  && key.length > 500) {
+    throw new Error('Failed to add annotation key: ' + key + ' value: ' + value + ' to subsegment ' +
+      this.name + '. Key must not be longer than 500 characters.');
+  } else if (typeof key === 'string' && !/^[A-Za-z0-9_]+$/.test(key)) {
+    throw new Error('Failed to add annotation key: ' + key + ' value: ' + value + ' to subsegment ' +
+      this.name + '. Key must be alphanumeric or underscores.');
```

Comment:
We generally do not do any regex checking on the SDK side due to performance. Based on some quick benchmarking this change could slow down adding an annotation anywhere from 7-90 times based on the length of the key. Also, similar to above, please log a warning and drop the annotation rather than throwing a runtime error if the key doesn't comply. 

An alternative is to try to match the key against an array of acceptable characters as is done in the [Python SDK](https://github.com/aws/aws-xray-sdk-python/blob/master/aws_xray_sdk/core/models/entity.py#L142). If this can be done in a succinct way that doesn't substantially impact performance, I'm open to a revision.